Feature/search and filter

ChawinTan opened the pull request:
## What's new

Added a search bar on top of the omni-panel for each item.

Added a new `SimpleSearch` component, a story for it as well as a unit test.

Dashboard:
![search_and_filtering](https://user-images.githubusercontent.com/29136440/111113027-1219b880-859c-11eb-8f64-64617736804e.png)

Story
![simple_search_story](https://user-images.githubusercontent.com/29136440/111113032-147c1280-859c-11eb-8f96-87aa2c6b9388.png)


<!-- NOTE: Pull request title should be "<package>: <summary>", if the PR affects multiple
  packages, use the main package that it affects. If the PR does not target any specific 
  packages, use general tags like "ci" or "versioning". -->

<!-- uncomment the next line if this PR fixes an issue -->
<!-- fixes #<issue-id> -->

<!-- Describe your changes.

  If your changes affects the UI, show screenshots or videos.

  If your changes affects, or is affected by other RMF components outside of this repo,
  describe how the components interact.

  If your changes fixes a bug, describe the root cause of the bug and how the
  proposed solution fixes it.

  If you went through several iterations while making this PR, explain why you
  prefer the proposed solution.
-->

## Self-checks

- [x] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [x] I added unit-tests for new components
- [x] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

## Discussion

<!-- Questions for reviewers, if any -->

koonpeng commented:
I think we can streamline it by removing the "X" button. An empty search term means show all items. We can also probably remove the `disabled` prop, I don't think we need it if the length is 0, are there any use case you can think of that uses it?

One thing we need to account for is spotlighting an item when a user click on a marker in the map. If the current filter does not include the item, what do you think is the best approach?
  * Do nothing, item won't be spotlighted
  * Automatically reset the filter
  * Show a prompt asking the user if they want to reset the filter

For the UI side, is it possible to make the separator between the search bar and the item a little more prominent?

ChawinTan commented:
> I don't think we need it if the length is 0, are there any use case you can think of that uses it?

My original intention was to disable the search bar if there are no equipment in that category, for example, there are no lifts/dispensers in this map, at least for now.

> One thing we need to account for is spotlighting an item when a user click on a marker in the map. If the current filter does not include the item, what do you think is the best approach

Thanks for the reminder, I think option 2 would suffice for this simple search. I feel that allowing user to select "no" for option 3 might not be needed because if the user clicks on an item, they most probably already want to see some details for that item and would mostly select "yes" to reset the search bar. At least that's what I think.

ChawinTan commented:
- Removed the X button and the disabled feature
- added a divider between search bar and accordion
- reset the search bar every time a new panel is selected
- reset the search bar and spotlight the item after the item is clicked
- updated wip-main-menu-plus-plus with the changes
![updated_search](https://user-images.githubusercontent.com/29136440/111253396-16061300-864e-11eb-8851-980673884611.png)

koonpeng commented on packages/dashboard/src/components/dashboard/dashboard.tsx:
```diff
@@ -308,55 +323,69 @@ export default function Dashboard(_props: {}): React.ReactElement {
-            ))}
+            <SimpleSearch onChange={onChange} value={search} />
+            {doors.map((door) => {
+              return door.name.includes(search) ? (
```
recommend for this to be case insensitive

koonpeng commented on packages/dashboard/src/components/dashboard/main-menu.tsx:
```diff
@@ -16,37 +16,44 @@ const useStyles = makeStyles((theme) => ({
 
 export interface MainMenuProps {
   pushView(view: OmniPanelViewIndex): void;
+  setSearch?: () => void;
```
I think we should have some docs for this prop.

On another note, this brings up some design consideration I have:

Should `MainMenu` manage the search state? Would it be better to expose on click handlers for each button and let `Dashboard` handle it? This also goes for `pushView`. I feel doing so is more "react-ive". The counter argument is that since there is no plan to use `MainMenu` outside of `Dashboard`, the usual react best practices does not offer any advantage, and doing it this way reduces the load/bloat on `Dashboard` and avoid it becoming somewhat of a "god class".

reference: https://reactjs.org/docs/thinking-in-react.html#step-4-identify-where-your-state-should-live

@matiasbavera Which do you think is the better approach?

koonpeng commented on packages/react-components/lib/simple-search.tsx:
```diff
@@ -0,0 +1,47 @@
+  },
+}));
+
+export const SimpleSearch = (props: SimpleSearchProps): JSX.Element => {
```
Do you think changing "Search" to "Filter" across the PR makes it more user friendly? Technically it is filtering instead of searching, having a text input with "Filter" might be better UX.

koonpeng reviewed (commented):
I think it's useful to add a e2e test to test the scenario where a filter is applied and user clicks on a marker.

ChawinTan commented:
I have left out the coverage of `dasbboard.tsx` for now. I have created a separate task for the mocking of schedule visualizer as that seem to be quite a big task on its own. I think we can also test for `setFilter` on the dashboard now since that part of the code is not affected, but the increment in coverage would be minimal and I feel that it is not worth it and would be better addressed when we successfully mock the `scheduleVisualizer`. For now, I have added some tests to `main-menu` to test for `pushViews` and `setFilters`.

- Added an end to end test
- updated Search to Filter
- updated wip-main-menu to keep in sync
- set filter term and equipment to lower caps to make it case-insensitive
- added test to main menu and wip main menu and updated it to use RTL.
- add some docs to `setFilter`

koonpeng commented on packages/dashboard/e2e/tests/filtering.test.ts:
```diff
@@ -0,0 +1,31 @@
+    door.click();
+
+    // check that the filter is empty after clicking
+    expect($('[data-component=simple-filter]').$('input').getValue()).toEqual('');
```
we should check if the door accordion for the clicked door is expanded as well.
